pkg/operator/certrotation/signer_test.go

```diff
@@ -19,6 +26,170 @@ import (
 	"github.com/openshift/library-go/pkg/operator/events"
 )
 
+func TestRotatedSigningCASecretWithMultipleControllers(t *testing.T) {
+	ns, name := "ns", "test-signer"
+	// represents a secret that was created before 4.7 and
+	// hasn't been updated until now (upgrade to 4.15)
+	existing := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:       ns,
+			Name:            name,
+			ResourceVersion: "10",
+			Annotations: map[string]string{
+				"auth.openshift.io/certificate-not-after":  "2108-09-08T22:47:31-07:00",
+				"auth.openshift.io/certificate-not-before": "2108-09-08T20:47:31-07:00",
+			},
+		},
+		Type: "SecretTypeTLS",
+		Data: map[string][]byte{"tls.crt": {}, "tls.key": {}},
+	}
+	if err := setSigningCertKeyPairSecret(existing, 24*time.Hour); err != nil {
+		t.Fatal(err)
+	}
+
+	// give it a second so we have a unique signer name,
+	// and also unique not-after, and not-before values
+	<-time.After(2 * time.Second)
+
+	// get the original crt and key bytes to compare later
+	tlsCertWant, ok := existing.Data["tls.crt"]
+	if !ok || len(tlsCertWant) == 0 {
+		t.Fatalf("missing data in 'tls.crt' key of Data: %#v", existing.Data)
+	}
+	tlsKeyWant, ok := existing.Data["tls.key"]
+	if !ok || len(tlsKeyWant) == 0 {
+		t.Fatalf("missing data in 'tls.key' key of Data: %#v", existing.Data)
+	}
+
+	// copy the existing object before test begins, so we can diff it against
+	// the final object on the cluster after the controllers finish
+	secretWant := existing.DeepCopy()
+
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	if err := indexer.Add(existing); err != nil {
+		t.Fatal(err)
+	}
+	clientset := kubefake.NewSimpleClientset(existing)
+
+	// the list cache is synced as soon as we have a delete, create, or update
+	clientset.PrependReactor("delete", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		if err = indexer.Delete(existing); err != nil {
+			t.Errorf("unexpected error while syncing the cache, op=delete: %v", err)
+		}
+		return false, nil, nil
+	})
+	clientset.PrependReactor("create", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.CreateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
+				t.Errorf("unexpected error while syncing the cache, op=create: %v", err)
+			}
+			return false, action.GetObject(), nil
+		}
+		t.Errorf("wrong test setup")
+		return false, nil, nil
+	})
+	clientset.PrependReactor("update", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.UpdateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
```

Comment:
Is this working around a problem with calling `Update` directly?

Reply:
yes, it ensures that the list cache always has the updated object